pkg/storage/tsdb/cacheable_index.go

```diff
@@ -0,0 +1,310 @@
+//
+// It uses as a key for the cache only the matcher and the name parameter.
+func (i *CacheableIndex) LabelValues(ctx context.Context, userID string, from, through model.Time, name string, matchers ...*labels.Matcher) ([]string, error) {
+	const opName = "LabelValues"
+
+	mountKeyFn := func(userID string, from, through model.Time, shard *index.ShardAnnotation, matcher *labels.Matcher) string {
+		if matcher == nil {
+			return name
+		}
+		return strings.Join([]string{name, matcher.String()}, sep)
+	}
+
+	fallbackFn := func(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matcher ...*labels.Matcher) ([][]byte, error) {
+		names, err := i.Index.LabelValues(ctx, userID, from, through, name, matchers...)
+		if err != nil {
+			return nil, errors.Wrap(err, "call to LabelValues")
+		}
+
+		var encodeBuf bytes.Buffer
+		enc := gob.NewEncoder(&encodeBuf)
+		if err := enc.Encode(names); err != nil {
+			return nil, errors.Wrap(err, "LabelValues encoding")
+		}
+
+		return [][]byte{encodeBuf.Bytes()}, nil
+	}
+
+	results, err := i.CacheableOp(ctx, opName, mountKeyFn, fallbackFn, userID, from, through, nil, matchers...)
+	if err != nil {
+		return nil, errors.Wrap(err, "cacheable index call to cacheable op LabelValues")
+	}
+
+	var values []string
+	var res []string
+	for _, val := range results {
+		decoderBuf := bytes.NewBuffer(val)
+		dec := gob.NewDecoder(decoderBuf)
+		if err := dec.Decode(&values); err != nil {
+			return nil, errors.Wrap(err, "cacheable index LabelValues decoding")
+		}
+
+		res = append(res, values...)
+	}
+
+	return res, nil
+}
+
+type mountKeyFunc func(userID string, from model.Time, through model.Time, shard *index.ShardAnnotation, matcher *labels.Matcher) string
+
+type fallbackFunc func(ctx context.Context, userID string, from model.Time, through model.Time, shard *index.ShardAnnotation, matcher ...*labels.Matcher) ([][]byte, error)
+
+// CacheableOp abstracts an operation that cache its results.
+//
+// It expects two generic functions: one to mount the key used to store and retrieve from cache and one that is invoked
+// whenever a cache miss occur, called fallbackFunc.
+// It uses the opName as a prefix key to insert in the cache to avoid collisions between different operations.
+func (i *CacheableIndex) CacheableOp(ctx context.Context, opName string, keyFn mountKeyFunc, fallbackFn fallbackFunc,
+	userID string, from, through model.Time, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([][]byte, error) {
+	var res [][]byte
+
+	// prepare a mapping of keys to matchers, used to retrieve matchers later.
+	keysMapping := make(map[string][]*labels.Matcher, len(matchers))
+	if len(matchers) == 0 {
+		key := keyFn(userID, from, through, shard, nil)
+		keysMapping[opName+sep+key] = []*labels.Matcher{}
+	}
+	for _, matcher := range matchers {
+		key := keyFn(userID, from, through, shard, matcher)
+		keysMapping[opName+sep+key] = []*labels.Matcher{matcher}
```

Comment:
I think we're being too clever here. Consider `{foo="bar", bazz="buzz"}`. This query asked for the _intersection_ of the two individual matchers.

If we have the following streams:
```
{foo="bar"}
{foo="bar", bazz="buzz"}
{foo="bar", bazz="buzz", bok="blob"}
{foo="bar", bazz="bonk"}
{bazz="buzz"}
```

the intersection will be
```
{foo="bar", bazz="buzz"}
{foo="bar", bazz="buzz", bok="blob"}
```

However, by requesting each matcher individually, we'd get the _union_ of the two:
```
{foo="bar"}
{foo="bar", bazz="buzz"}
{foo="bar", bazz="buzz", bok="blob"}
{bazz="buzz"}
```

That is not what we want. Instead, we should make one call to the cache for `{foo="bar", bazz="buzz"}`.

Reply:
Good catch, thanks. I'll be looking for a way to unify all matchers into a single string then.

Reply:
pushed a commit that fixes this (100% based on our pairing today, thanks! :smile: )